url.c

```diff
@@ -96,20 +96,76 @@ enum UrlScheme url_check_scheme(const char *s)
     return (enum UrlScheme) i;
 }
 
+
+static int parse_query_string(struct Url *u, char *src)
+{
+  struct UrlQueryString *qs = NULL;
+  char *k = NULL, *v = NULL;
+
+  while (src && *src)
+  {
+    qs = safe_calloc(1, sizeof(struct UrlQueryString));
+    if ((k = strchr(src, '&')))
+      *k = '\0';
+
+    if ((v = strchr(src, '=')))
+    {
+      *v = '\0';
+      qs->value = v + 1;
+      if (url_pct_decode(qs->value) < 0)
+      {
+        FREE(&qs);
+        return -1;
+      }
+    }
+    qs->name = src;
+    if (url_pct_decode(qs->name) < 0)
+    {
+      FREE(&qs);
+      return -1;
+    }
+    STAILQ_INSERT_TAIL(&u->query_strings, qs, entries);
+
+    if (!k)
+      break;
+    src = k + 1;
+  }
+  return 0;
+}
+
+
 /**
- * parse_userhost - fill in components of Url with info from src
+ * url_parse - Fill in Url
+ *
+ * @param u        struct Url where the result of stored
+ * @param src      pointer to the string to parse
+ * @param parse_qs if we should parse the query string
+ * @retval 0       if src is valid
+ * @retval -1      if src is invalid
+ *
+ * char* elements are pointers into src, which is modified by this call
+ * (duplicate it first if you need to).
+ *
+ * This method doesn't allocated any additional char* of the Url and
+ * UrlQueryString structs.
+ *
+ * To free Url, caller must free "src" and call url_free(&u);
```

Comment:
Perhaps `call url_free(u)` since `u` is already a pointer. It's a minor thingy, but this works a bit different than most other free function in NeoMutt, where you pass in a pointer-to-pointer and the free function also sets the pointer to NULL at the caller.